### ISSUE TITLE
FR: Make databaseAuthVariableOverride work for Firestore

### ISSUE THREAD

**Issue by adamduren** [firebase/firebase-admin-node#132]
From what I understand databaseAuthVariableOverride does not work for Firestore. This was important in the real time database to make sure that scripts don't execute with unchecked permissions. Is there an alternative for Firestore?

**Comment by hiranya911** [comment]
I believe there's some ongoing work to support this. @mikelehen should know more.

**Comment by mikelehen** [comment]
cc/ @mcdonamp 

This is a good feature request and we've had some internal discussions about how we might try to support this in the future, but it will probably not be something we can accommodate in the near term unfortunately.

**Comment by mcdonamp** [comment]
The near term recommendation that I offer is: have your function [mint a custom auth token](https://firebase.google.com/docs/auth/admin/create-custom-tokens) and then use the client libs instead of the admin ones.

In the future, we're seeing how we could either:

- Bake this functionality into the SDK
- Add a side-channel to accept a second credential (the user credential) in addition to the service credential

**Comment by adamduren** [comment]
Yea, that was the first thought I had but the js-sdk is incompatible with nodeper  https://github.com/firebase/firebase-js-sdk/issues/221.

**Comment by mikelehen** [comment]
@adamduren You are correct.  We may have a solution for that in the next week or two though. Stay tuned.

**Comment by degroote22** [comment]
I'd love to be able to continue to use my existing validation rules on the server.
It really helps me on keeping the database consistent and monomorphic...
I guess that validation based only on the data format is not possible without also taking into account authentication rules and permissions, right?
Do you guys have any news to share with us? Thanks.

**Comment by mcdonamp** [comment]
@degroote22 this is still a topic in active discussion, see https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/google-cloud-firestore-discuss/GYROSnxlKro/XZeejWfPAwAJ for some more thoughts.

**Comment by kglowacki** [comment]
+1
I consider this to be a key security feature... any update?

**Comment by ElePhontitis** [comment]
Can we expect this in the near future? It would be really helpful!

**Comment by mcdonamp** [comment]
Unfortunately it requires some changes to underlying infrastructure that I don't see coming anytime soon, so I'd rely on also performing server-side validation when using those clients.

**Comment by lookfirst** [comment]
Given that the transaction API of the client sdk doesn't match the transaction api of the admin sdk, I can't even replace the sdk usage in my functions with the client sdk. =(

https://github.com/firebase/firebase-js-sdk/issues/2112

**Comment by mesqueeb** [comment]
I believe that I hit a wall until I found this issue.
My problem is outlined here:
https://stackoverflow.com/questions/63294542/cannot-impersonate-user-with-firebase-admin-during-firestore-calls

I believe the answer is: **it's not possible, because Google didn't make impersonating possible for Firestore?**

Is my assumption correct?

**Comment by schmidt-sebastian** [comment]
That statement is still correct, but this is an often requested feature, so don't give up hope yet :)

**Comment by jakebiesinger-storyhealth** [comment]
For integrity and security Reasons, we use Cloud Functions as an API layer and never directly write to Firestore from the client. We're looking at how to better manage access control and this feature would make a huge difference. 

Specifically, we could use the authVariableOverride to scope Admin API requests (from our API layer) down to the user calling the function, which would in turn allow us to use native security rules instead of having to come up with our own access control system.

I hope you consider reprioritizing this functionality. It would really improve the flexibility of the Admin API and provide improved security at the same time.

**Comment by magnusnilsen** [comment]
Would love this feature! Any updates?

**Comment by abba-logico-srl** [comment]
upvoting this feature request here too 👍 
Any updates?

**Comment by sanny-io** [comment]
I'd love to have this.

**Comment by saidgeek** [comment]
+1 to this feature, is very important for can use a native rules, any update?

**Comment by SamyPesse** [comment]
@schmidt-sebastian any update from the Firebase team?

It looks like a highly requested feature, especially for large applications built on Firebase. At the moment we are using the alternative of authenticating the `firebase-js-sdk` with a custom token, but at scale, we can easily reach some quota on `identitytoolkit.googleapis.com`.
